src/db/spanner/models.rs

```diff
@@ -641,95 +688,204 @@ impl SpannerDb {
+        sql.params = Some(sqlparams);
+
+        let results = spanner
+            .hub
+            .projects()
+            .instances_databases_sessions_execute_sql(sql, session)
+            .doit();
+        match results {
+            Ok(_results) => {}
+            // TODO Return the correct error
+            Err(_e) => return Err(DbErrorKind::CollectionNotFound.into()),
         }
+        self.get_storage_timestamp_sync(params.user_id)
     }
 
     pub(super) fn touch_collection(
         &self,
         user_id: u32,
         collection_id: i32,
     ) -> Result<SyncTimestamp> {
+        // XXX: We should be able to use Spanner's insert_or_update here (unlike w/ put_bsos)
         let spanner = &self.conn;
         let session = spanner.session.name.as_ref().unwrap();
-        let mut sql = ExecuteSqlRequest::default();
-        sql.sql = Some("INSERT INTO user_collections (user_id, collection_id, modified) VALUES (@userid, @collectionid, @modified) ON DUPLICATE KEY UPDATE modified = @modified".to_string());
+        let mut sql = self.sql_request("SELECT 1 as count FROM user_collections WHERE userid = @userid AND collection = @collectionid;")?;
         let mut sqlparams = HashMap::new();
         sqlparams.insert("userid".to_string(), user_id.to_string());
         sqlparams.insert("collectionid".to_string(), collection_id.to_string());
         sql.params = Some(sqlparams);
-
         let results = spanner
             .hub
             .projects()
             .instances_databases_sessions_execute_sql(sql, session)
             .doit();
-        match results {
-            Ok(_results) => Ok(self.timestamp()),
-            // TODO Return the correct error
-            Err(_e) => Err(DbErrorKind::CollectionNotFound.into()),
+        let exists = match results {
+            Ok(results) => results.1.rows.is_some(),
+            _ => return Err(DbErrorKind::CollectionNotFound.into()),
+        };
+
+        if exists {
+            let mut sql = self.sql_request("UPDATE user_collections SET last_modified=@last_modified WHERE userid=@userid AND collection=@collectionid")?;
+            let mut sqlparams = HashMap::new();
+            let mut sqltypes = HashMap::new();
+            sqlparams.insert("userid".to_string(), user_id.to_string());
+            sqlparams.insert("collectionid".to_string(), collection_id.to_string());
+            let timestamp = self.timestamp().as_i64();
+            let modifiedstring = to_rfc3339(timestamp)?;
+            sqlparams.insert("last_modified".to_string(), modifiedstring);
+            sqltypes.insert(
+                "last_modified".to_string(),
+                Type {
+                    array_element_type: None,
+                    code: Some("TIMESTAMP".to_string()),
+                    struct_type: None,
+                },
+            );
+            sql.params = Some(sqlparams);
+            sql.param_types = Some(sqltypes);
+
+            let results = spanner
```

Comment:
could probably collapse this and the following match and remove the `let results`.